switchtec.c

```diff
@@ -1549,6 +1549,25 @@ static int switchtec_init_pci(struct switchtec_dev *stdev,
 	return 0;
 }
 
+static enum switchtec_rel get_release(struct switchtec_dev *stdev)
+{
+	u32 fw_ver = ioread32(&stdev->mmio_sys_info->firmware_version);
+
+	if (stdev->gen == SWITCHTEC_GEN3) {
+		if ((stdev->var == SWITCHTEC_PFX ||
+		     stdev->var == SWITCHTEC_PSX) &&
+		    fw_ver <= 0x10b0080)
+			return SWITCHTEC_REL_GEN3_PFX_PSX_MR4;
+		else if (stdev->var == SWITCHTEC_PAX &&
+			 fw_ver <= 0x2060032)
+			return SWITCHTEC_REL_GEN3_PAX;
+	} else if (stdev->gen == SWITCHTEC_GEN4) {
+		return SWITCHTEC_REL_GEN4;
+	}
+
+	return SWITCHTEC_REL_UNKNOWN;
+}
+
```

Comment:
This doesn't make any sense to me. So if we release new firmware versions for PFX/PSX or PAX they well start being SWITCHTEC_REL_UNKNOWN and things will break in unknown ways? That's nuts. 

Based on this it seems like you should just be using the 'var' field directly.

Reply:
Also, the release variable was never used in subsequent patches, so it looks like you can just drop this patch.

Reply:
I should have more patches in this series to show the way I wanted to use the 'rel' field. I wanted to check 'rel' together with other versions to test if a device can be supported and exit with error if found it's unsupported in probe(). So with the 'SWITCHTEC_REL_UNKNOWN', the driver will choose to quit in probe(), just like we failed init isr.

Generally, we have three types of unknown situations. 

1. unknown devices (gen/variable)
For this case, I think it's quite straight forward, it's not in our pci device table.
2. unknown releases for either gen3 or gen4
I'd like to know your comment about what we should do in this situation.
For gen3, current latest is MR4, if we have MR5 in the future, should current driver be loaded or just exit with error in the probe(). We have no extra versions in gen3, so we have no idea what are changed in MR5.
For gen4, release is not that important, as we have other versions for similar reason.
3. unknown versions for gen4.  
We have the following versions since gen4. 

a, GAS layout version(major.minor). 
Compatible changes in GAS definition trigger a minor increase, while incompatible ones trigger major change.
b, Management cmd set version(major.minor) 
Compatible changes in any MRPC cmd, trigger a minor increase, while incompatible ones trigger a major increase.
c, Fabric cmd set version(major.minor) 
Compatible changes in any MRPC cmd, trigger a minor increase, while incompatible ones trigger a major increase.
d, MRPC UART protocol version.
Any update in this protocol triggers version change.
e, MRPC TWI protocol version
Any update in this protocol triggers version change.
f, MRPC Ethernet Protocol version
Any update in this protocol triggers version change.
g, MRPC inbadn Protocol Version
Any update in this protocol triggers version change. For switchtec driver, for the devices of generation greater then gen3, I think we need to quit with error in probe() on this cases.

Reply:
That's a big NAK on quiting probe if the release is unknown. This makes the driver not work with newer firmware. No. No. No.

If new firmware breaks something that's on them, they shouldn't do that. 

The driver supports all the devices in the vendor/product ID table. Us and the firmware team make sure released firmware works with all the devices specified in that table. We aren't blessing specific firmware versions to work with the driver and breaking with anything else. 

The only reason to have a release/firmware version variable is to fix a nasty bug and that means the patch that adds the release variable must be accompanied by the patch that fixes the bug and we must be in a tight spot that we have to make it depend on the firmware version.

I don't care about hypothetical changes, we will deal with them when we see them. I care about specific issues we need to deal with right now, and the only one you've put forward so far is the PAX vs non-PAX events which doesn't seem that big an issue and only means we need to know PAX vs not-PAX and not specific firmware versions.

Reply:
Ok. please check the PR #5.